### PR TITLE
[NOSQUASH] Resync from Apache

### LINKED ISSUE #244: Merge SPARK-21447
We hit this issue for a lot of super-large spark applications (30k+ jobs), currently bouncing it was the only fix

https://github.com/apache/spark/pull/18673

### PR DESCRIPTION
**Pull request by ash211**
Fixes #244

### CODE REVIEW

**Comment by ash211** [comment]
Hmmm looks like we might have a problem with bumping netty.  We took it from the 4.0.x series to the 4.1.x series in https://github.com/palantir/spark/pull/179/files#diff-600376dffeb79835ede4a0b285078036R626

But now when pulling in latest Apache code, I'm seeing this failure:

```
sbt.ForkMain$ForkError: sbt.ForkMain$ForkError: java.lang.NoClassDefFoundError: io/netty/util/collection/IntObjectMap$Entry
	at org.apache.arrow.vector.util.MapWithOrdinal$1$1.apply(MapWithOrdinal.java:136)

```

Arrow (0.4.0) depends on [netty 4.0.41](https://github.com/apache/arrow/blob/apache-arrow-0.4.0/java/pom.xml#L457)

And between netty 4.0.43 (current Spark master version) and 4.1.13 (current Palantir master version) that "entry" class was renamed:

```
aash@aash01-mac ~$ jar tf /Users/aash/.gradle//caches/modules-2/files-2.1/io.netty/netty-all/4.0.43.Final/9781746a179070e886e1fb4b1971a6bbf02061a4/netty-all-4.0.43.Final.jar | grep IntObjectMap
io/netty/util/collection/IntObjectMap.class
io/netty/util/collection/IntObjectMap$Entry.class
aash@aash01-mac ~$ jar tf /Users/aash/.gradle//caches/modules-2/files-2.1/io.netty/netty-all/4.1.13.Final/e1f2d1ee50626697d3ea73a11f1c7301e8589ef8/netty-all-4.1.13.Final.jar | grep IntObjectMap
io/netty/util/collection/IntObjectMap$PrimitiveEntry.class
io/netty/util/collection/IntObjectMap.class
aash@aash01-mac ~$
```

Possibly we need to drop back down to the netty 4.0 series until Spark/Arrow go to netty 4.1 ?

cc @robert3005 @mccheah

**Comment by ash211** [comment]
Rename happened here, only in the 4.1.x line: https://github.com/netty/netty/commit/93fc3c6e45cda454ba74e6430a8b7b22bdf1ee15#diff-dd65461359cfebaaae8d8c1590a5ffa9R31

**Comment by robert3005** [comment]
Sucks. In light of https://issues.apache.org/jira/browse/SPARK-19552 we will have to downgrade.

**Comment by ash211** [comment]
Bummer -- downgraded at https://github.com/palantir/spark/pull/247/commits/bd089349c65eb586f238a02d4cad460d30935771

**Comment by ash211** [comment]
Ready for review @mccheah / @robert3005 .  The last few commits are the interesting ones.

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java [hunk 1]
```diff
@@ -206,21 +206,11 @@ public UnsafeRow next() {
 
     @Override
     public int compare(Object baseObj1, long baseOff1, Object baseObj2, long baseOff2) {
-<<<<<<< HEAD
```
Comment (mccheah) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java]: Hm, what was the context for these conflicts?

Comment (ash211) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java]: We merged a PR into Palantir Spark at https://github.com/palantir/spark/pull/222 but the final version later submitted upstream at https://github.com/apache/spark/pull/18543 and reworked into merged PR https://github.com/apache/spark/pull/18679 came out differently.  This resolves the conflict between the two with the version merged upstream in 18679

**Pull request merged**